In [1]:
import pandas as pd
import numpy as np
import math


In [6]:
df = pd.read_csv("household_2020a.csv")
df.rename(columns={"income_d":"income"},inplace=True)
df

year metro         id  sex  age  number  education  marriage   asset  \
0      2020    G1   10000112    1   34       3          6         2  112000   
1      2020    G1   10000132    2   45       2          8         2   42500   
2      2020    G1   10000162    2   73       1          2         3    5712   
3      2020    G1   10000182    1   58       2          4         2   14870   
4      2020    G1   10000192    2   27       1          4         1     814   
...     ...   ...        ...  ...  ...     ...        ...       ...     ...   
18059  2020    G2  993800291    1   42       3          6         2    6705   
18060  2020    G2  994800251    1   49       5          6         2  130180   
18061  2020    G2  994800261    1   57       5          6         2  265226   
18062  2020    G2  994800291    2   53       2          7         4  109695   
18063  2020    G2  995800261    1   44       5          6         2   71223   

        debt  income industry  job  house  
0      54500    4599        F    3      2  
1      17500   15257        J    2      3  
2          0     725        T    4      2  
3          0    2431        C    5      2  
4          0     893        R    2      3  
...      ...     ...      ...  ...    ...  
18059   2200    -390      NaN  NaN      3  
18060  22000   11784        O    3      1  
18061  29000   11560      NaN  NaN      1  
18062  25000    7074        O    2      2  
18063  13000    5404        C    3      1  

[18064 rows x 14 columns]

In [7]:
df.to_csv('household_2020a.csv')

In [8]:
df = pd.read_csv("household_2020a.csv")
df

Unnamed: 0  year metro         id  sex  age  number  education  \
0               0  2020    G1   10000112    1   34       3          6   
1               1  2020    G1   10000132    2   45       2          8   
2               2  2020    G1   10000162    2   73       1          2   
3               3  2020    G1   10000182    1   58       2          4   
4               4  2020    G1   10000192    2   27       1          4   
...           ...   ...   ...        ...  ...  ...     ...        ...   
18059       18059  2020    G2  993800291    1   42       3          6   
18060       18060  2020    G2  994800251    1   49       5          6   
18061       18061  2020    G2  994800261    1   57       5          6   
18062       18062  2020    G2  994800291    2   53       2          7   
18063       18063  2020    G2  995800261    1   44       5          6   

       marriage   asset   debt  income industry  job  house  
0             2  112000  54500    4599        F    3      2  
1             2   42500  17500   15257        J    2      3  
2             3    5712      0     725        T    4      2  
3             2   14870      0    2431        C    5      2  
4             1     814      0     893        R    2      3  
...         ...     ...    ...     ...      ...  ...    ...  
18059         2    6705   2200    -390      NaN  NaN      3  
18060         2  130180  22000   11784        O    3      1  
18061         2  265226  29000   11560      NaN  NaN      1  
18062         4  109695  25000    7074        O    2      2  
18063         2   71223  13000    5404        C    3      1  

[18064 rows x 15 columns]

In [9]:
df1 = df[['income','education']]

In [10]:
def edu(x):  # 교육코드를 교육연수로 바꾸어주는 함수
    if x == 2:
        return 6
    elif x == 3:
        return 9
    elif x == 4:
        return 12
    elif x == 5:
        return 14
    elif x == 6: 
        return 16
    elif x == 7:
        return 18
    elif x == 8:
        return 21
    else:
        return 0

In [14]:
df2 = df1.copy()
df2['education_year'] = df2['education'].apply(edu) # apply 함수 사용
df3 = df2.drop(['education'],axis=1)
corr = df3.corr()
corr

income  education_year
income          1.000000        0.390346
education_year  0.390346        1.000000

In [15]:
np.round(corr.iloc[0, 1], 3)

0.39

### 상관계수 함수를 만들기. 상관계수를 구하는 데는 lamda,map 함수를 이용한다. 람다 함수는 한줄 함수로 함수를 간단하게 표현할 수 있다.

In [18]:

def squarel(a):
    return a*a
squarel(5)



25

In [19]:
square2 = lambda a: a*a
square2(5)

25

### map 함수는 map(function,iterable) 형태를 가지고 있으며, iterable에는 리스트나 튜플이 올 수 있습니다. map 함수는 객체 형태이므로 list나 tuple로 변환시켜 사용한다. 예를 들어, 리스트 형태인 [1,2,3]을 각각 제곱하여 반환하는 함수를 map함수와 lambda 함수를 이용하여 다음과 같이 간단하게 표현할 수 있다.

In [20]:
mylist = [1,2,3]
result1 = []
for i in mylist:
    i2 = squarel(i)
    result1.append(i2)
result1


[1, 4, 9]

In [21]:
result2 = list(map(lambda a: a*a,[1,2,3]))
result2

[1, 4, 9]

### 다음은 복수의 함수를 이용하여 상관계수를 구하는 과정이다. 분산을 구하는 함수에 필요한 제곱항의 합을 구하는 과정에는 lambda 함수와 map 함수를 이용한다.

In [25]:
def de_mean(x):
    x_bar = np.mean(x)
    return[x_i-x_bar for x_i in x] # list comprehensio,for 반복문을 사용하여 list를 효과적으로 생성
def covariance(x,y):  # 공분산
    n = len(x)
    return np.dot(de_mean(x),de_mean(y))/(n-1)
def sum_of_squares(x):
    return sum(list(map(lambda a : a*a,x)))
def variance(x):
    n = len(x)
    deviation = de_mean(x)
    return sum_of_squares(deviation)/(n-1)
def standard_deviation(x):
    return math.sqrt(variance(x))
def correlation(x,y):
    stdev_x = standard_deviation(x)
    stdev_y = standard_deviation(y)
    return np.round(covariance(x,y)/(stdev_x*stdev_y),2)

In [26]:
edu = df3.to_numpy()
x = edu[:,0]
y = edu[:,1]

In [27]:
correlation(x,y)

0.39

### 30대 1인 가구의 소득과 자산 간 회귀분석


In [28]:
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
from matplotlib import rcParams


In [30]:
from matplotlib import font_manager,rc
## 그래프에서 한글 폰트 이용하기
rcParams['axes.unicode_minus'] = False
plt.rcParams['font.family'] = 'Malgun Gothic'
font_name = font_manager.FontProperties(fname="c:/windows/Fonts/malgun.ttf").get_name()
rc('font',family = font_name)

In [31]:
df = pd.read_csv("household_2020a.csv")
df['income'] = df['income']/10000  # 억 원으로 단위조정
df['net_asset'] = df['asset']/10000 - df['debt']/10000  # 순자산
df1 = df.loc[df['number'].isin([1])&(df['age']>=30) & (df['age']<39)] # 30대 1인 가구
df1 = df1[['income','net_asset']]

### statsmodels 패키지를 이용하여 회귀분석을 해보자. 종속변수는 순자산이고 설명변수는 소득이다.

In [32]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
asset_model = ols("net_asset ~ income",data = df1).fit()
print(asset_model.summary())

                            OLS Regression Results                            
Dep. Variable:              net_asset   R-squared:                       0.105
Model:                            OLS   Adj. R-squared:                  0.103
Method:                 Least Squares   F-statistic:                     41.24
Date:                Tue, 28 Mar 2023   Prob (F-statistic):           4.39e-10
Time:                        17:08:30   Log-Likelihood:                -596.66
No. Observations:                 352   AIC:                             1197.
Df Residuals:                     350   BIC:                             1205.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3232      0.135      2.389      0.0

### RMSE 구하기
### 추정오차를 구하기 위해 회귀직선으로 추정한 추정치를 구하고 실제값과 추정치와의 각각의 오차(er)를 구한 다음 각각의 오차를 이용하여 RMSE를 구한다.

In [33]:
df2 = df1.copy()
df2['predicted'] = asset_model.params[0] + df2['income']*asset_model.params[1]
df2['er'] = df2['net_asset'] - df2['predicted']
df2.head()

income  net_asset  predicted        er
19   0.1033     0.5252   0.586148 -0.060948
134  0.3032     1.2920   1.095055  0.196945
157  0.1812     0.0540   0.784467 -0.730467
190  0.3959    -0.4780   1.331052 -1.809052
191  0.1197     0.1610   0.627900 -0.466900

In [37]:
error = df2['er'].to_numpy()
error2 = []
for i in range(len(error)):
    er2 = error[i] ** 2
    error2.append(er2)
rmse = np.sqrt(np.sum(error2)/(len(error)-2))
np.round(rmse,2)

1.32

### 소득이 5천만 원인 30대 1인 가구가 순자산이 2억 원 이상 확률


In [38]:
import scipy.stats as st
intercept = asset_model.params[0]
slope = asset_model.params[1]
asset1_mn = intercept + 0.5*slope
dif = 2 - asset1_mn
z = dif/rmse
np.round(z,3)


0.306

In [39]:
p_value = 1 - st.norm.cdf(z)
np.round(p_value,2)


0.38

### 결정계수 
### SST,SSR,SSE를 각각 구한 후 결정계수와 조정된 결정계수를 구합니다.


In [42]:
df3 = df2.copy()
df3['y_bar'] = df3['net_asset'].mean()
df3['sst'] = (df3['net_asset']-df3['y_bar'])**2
df3['ssr'] = (df3['predicted']-df3['y_bar'])**2
df3['sse'] = (df3['net_asset']-df3['predicted'])**2
df4=df3[['sst','ssr','sse']]
df5 = df4.sum(axis = 0)
df5

sst    683.514496
ssr     72.051564
sse    611.462932
dtype: float64

In [43]:
r2 = df5['ssr']/df5['sst']
np.round(r2,3)

0.105

In [44]:
k = 2
n = len(df1)
a = df5['sse']/(n-k-1)
b = df5['sst']/(n-1)
adjR2 = 1 - a/b
np.round(adjR2,2)

0.1

# 2 try

In [1]:
import pandas as pd
import numpy as np
import math


In [4]:
df = pd.read_csv("household_2020a.csv")
df1 = df[['income','education']]
df1

income  education
0        4599          6
1       15257          8
2         725          2
3        2431          4
4         893          4
...       ...        ...
18059    -390          6
18060   11784          6
18061   11560          6
18062    7074          7
18063    5404          6

[18064 rows x 2 columns]

In [6]:
def edu(x):
    if x == 2:
        return 6
    elif x == 3:
        return 9
    elif x == 4:
        return 12
    elif x == 5:
        return 14
    elif x == 6:
        return 16
    elif x == 7:
        return 18
    elif x == 8:
        return 21
    else:
        return 0
    

In [7]:
df2 = df1.copy()
df2['education_year'] = df2['education'].apply(edu) # apply 함수 사용 = ()안에 있는 함수에 df2[] 안에 있는걸 적용
df3 = df2.drop(['education'],axis=1) #drop은 해당 열을 삭제함
corr = df3.corr()  #corr()은 누락값을 제외하고 전체(dataframe) 컬럼들 간의 상관도를 계산한다.
corr

income  education_year
income          1.000000        0.390346
education_year  0.390346        1.000000

In [8]:
np.round(corr.iloc[0,1],3)       ????

0.39

In [9]:
(lambda a:a*a)(5)

25

In [11]:
myList = [1,2,3]
result1 = []
for i in myList:
    i2 = (lambda a:a*a)(i)
    result1.append(i2)
result1

[1, 4, 9]

In [12]:
result2 = list(map(lambda a: a*a,[1,2,3]))
result2

[1, 4, 9]

In [13]:
def de_mean(x):
    x_bar = np.mean(x)
    return[x_i - x_bar for x_i in x]
def covariance(x,y):   # 공분산
    n = len(x)
    return np.dot(de_mean(x),de_mean(y))/(n-1)
def sum_of_squares(x):
    return sum(list(map(lambda a : a*a,x)))
def variance(x):
    n = len(x)
    deviation = de_mean(x)
    return sum_of_squares(deviation)/(n-1)
def standard_deviation(x):
    return math.sqrt(variance(x))   # sqrt 제곱근 구하기
def correlation(x,y):
    stdev_x = standard_deviation(x)
    stdev_y = standard_deviation(y)
    return np.round(covariance(x,y)/(stdev_x*stdev_y),2)      # 넘파이 반올리 뒤에 숫자는 자릿수
edu = df3.to_numpy()
x = edu[:,0]
y = edu[:,1]
correlation(x,y)


0.39

# 7.7 소득과 자산 간 회귀분석

In [16]:
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
from matplotlib import rcParams

In [19]:
from matplotlib import font_manager,rc
rcParams['axes.unicode_minus']=False
plt.rcParams['font.family'] = 'Malgun Gothic'
font_name = font_manager.FontProperties(fname="c:/Windows/Fonts/malgun.ttf").get_name()
rc('font,family = font_name')

In [22]:
df = pd.read_csv("household_2020a.csv")
df['income'] = df['income']/10000
df['net_asset'] = df['asset']/10000 - df['debt']/10000
df1 = df.loc[df['number'].isin([1])&(df['age'] >= 30)&(df['age']<39)]   # isin = 일치하는지
df1 = df1[['income','net_asset']]

# statsmodels 패키지를 이용하여 회귀분석. 
## 종속변수는 순자산(net asset)이고 설명변수는 소득(income)이다. 절편 = 0.3232,기울기 = 2.5458

In [23]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
asset_model = ols("net_asset ~ income",data = df1).fit()
print(asset_model.summary())

                            OLS Regression Results                            
Dep. Variable:              net_asset   R-squared:                       0.105
Model:                            OLS   Adj. R-squared:                  0.103
Method:                 Least Squares   F-statistic:                     41.24
Date:                Sat, 01 Apr 2023   Prob (F-statistic):           4.39e-10
Time:                        16:49:32   Log-Likelihood:                -596.66
No. Observations:                 352   AIC:                             1197.
Df Residuals:                     350   BIC:                             1205.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3232      0.135      2.389      0.0

## RMSE 구하기

In [29]:
df2 = df1.copy()
df2['predicted'] = asset_model.params[0] + df2['income']*asset_model.params[1]
df2['er'] = df2['net_asset']-df2['predicted']
df2.head()

income  net_asset  predicted        er
19   0.1033     0.5252   0.586148 -0.060948
134  0.3032     1.2920   1.095055  0.196945
157  0.1812     0.0540   0.784467 -0.730467
190  0.3959    -0.4780   1.331052 -1.809052
191  0.1197     0.1610   0.627900 -0.466900

In [31]:
error = df2["er"].to_numpy()
error2 = []
for i in range(len(error)):
    er2 = error[i] ** 2
    error2.append(er2)
    
rmse = np.sqrt(np.sum(error2)/(len(error)-2))
np.round(rmse,2)


1.32

## 소득이 5처만 원인 30대 1인 가구가 순자산이 2억 원 이상 확률


In [32]:
import scipy.stats as st
intercept = asset_model.params[0]
slope = asset_model.params[1]
asset1_mn = intercept + 0.5*slope
dif = 2-asset1_mn
z = dif/rmse
np.round(z,3)


0.306

In [33]:
p_value = 1-st.norm.cdf(z)
np.round(p_value,2)

0.38

# 결정계수
## SST,SSR,SSE를 각각 구한 후 결정계수와 조정된 결정계수를 구합니다.

In [34]:
df3 = df2.copy()
df3['y_bar'] = df3['net_asset'].mean()
df3['sst'] = (df3['net_asset']-df3['y_bar'])**2
df3['ssr'] = (df3['predicted']-df3['y_bar'])**2
df3['sse'] = (df3['net_asset']-df3['predicted'])**2
df4=df3[['sst','ssr','sse']]
df5 = df4.sum(axis=0)
df5



sst    683.514496
ssr     72.051564
sse    611.462932
dtype: float64

In [35]:
r2 = df5['ssr']/df5['sst']
np.round(r2,3)

0.105

In [36]:
k=2
n = len(df1)
a = df5['sse']/(n-k-1)
b = df5['sst']/(n-1)
adjR2 = 1- a/b
np.round(adjR2,2)

0.1